### DISCUSSION TITLE
Archiving null-ls

### DISCUSSION THREAD

**Discussion by skinner12** [LazyVim/LazyVim#1080] [General]
[null-ls](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/1621) will be archived.

**Comment by LintaoAmons** [comment]
Wondering what is his main editor right now

**Comment by fredrikaverpil** [comment]
There's an ongoing discussion over at Reddit about what to use instead: https://www.reddit.com/r/neovim/comments/14v6oet/nullls_will_be_archived/?sort=new

Since LazyVim uses null-ls extensively, I'll wait and see what @folke believes is the way forward for the distro and I'll likely follow suit in my own config, building on top of that.

**Comment by folke** [comment]
I'll wait a bit to see if someone's willing to pick up a fork of null-ls.

For me personally, I'm mostly interested in the formatters and there's other plugins that can fulfill that role.

The only other part of null-ls I'm interested in is the diagnostics. Would actually be better if a separate plugin took care of that.

For both functionalities, tying in to the LSP framework, is not really needed, so would actually be better if that would no longer be implemented like it is right now in null-ls.

Let's see what happens over the next couple of weeks.

**Reply by realeinherjar** [reply]
Seem that [`mhartington/formatter.nvim`](https://github.com/mhartington/formatter.nvim) will be the chosen one.
It's been almost 2 monthts, any thoughts on that?

**Comment by fredrikaverpil** [comment]
> For me personally, I'm mostly interested in the formatters and there's other plugins that can fulfill that role.

Do you have any particular one in mind?
Maybe https://github.com/mhartington/formatter.nvim ?

> The only other part of null-ls I'm interested in is the diagnostics. 

I just dabbled around with https://github.com/mfussenegger/nvim-lint and it works pretty good I think. I like how it falls back to Mason-installed binaries in case none is found on `$PATH`.

> Let's see what happens over the next couple of weeks.

Yes, no need to panic 😄 btw, I noticed Jose is also archiving typescript.nvim: https://github.com/jose-elias-alvarez/typescript.nvim/issues/80

**Reply by folke** [reply]
I'm not that worried about typescript.nvim.

Most functionality is actually already possible in Neovim.

To get the source actions in LazyVim, use `<leader>cA`, so capital `A`. Those are coming from tsserver itself.

**Comment by danielo515** [comment]
The functionality null-ls provides around easyness of adding should not be understated.
I also use the code actions helpers a lot, so I am indeed worried about this

**Reply by fredrikaverpil** [reply]
FWIW, I played around with replacing null-ls, although I've got these replacements disabled for now. I'd like to stick with whatever LazyVim will use so to have its `<leader>uf` (toggle formatting on save) still working. Figured I'd just share this here for anyone who might be interested.

Formatting:
- https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/format.lua

Linting:
- https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/lint.lua
- https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/config/autocmds.lua#L38

I guess I primarily rely on code actions from LSPs rather than linters, so I haven't really looked into how `mfussenegger/nvim-lint` deals (or doesn't deal) with that.

**Reply by lotus-x** [reply]
Only thing I couldn’t find an alternative is the code actions. Actually I’m using cspell and eslint code action helpers with null-ls, but even if eslint can be replaced by the lsp server I can’t find a solution for cspell.

I was going through both efm and nvim-lint issues, mfussenegger has no plans to support dynamic code actions as it seems, https://github.com/mfussenegger/nvim-lint/issues/5
Efm has not yet being implemented the code action support.

**Comment by directormac** [comment]
The null_ls archive got me worried and tried using vscode again. 🤸

**Comment by elentok** [comment]
I've been using null-ls for a long time for both formatting and linting.

For linting I replaced it with nvim-lint which seems to work pretty well so far.

For formatting I had my own implementation for format-on-save that only used LSP formatting and non-LSP via null-ls, now that null-ls is archived I added direct support for formatting using non-LSP formatters and extracted the whole thing as a plugin - https://github.com/elentok/format-on-save.nvim

I tried to keep it as simple as possible:

- **Synchronous formatting** - async formatting is very tricky since you need to block the buffer from editing or deal with timing issues where the buffer has been edited formatting. I tried using formatter.nvim but couldn't get it to play well with LSP and the async thing didn't work well for me.
- **Format on save only** - No partial formatting, only format entire files (supporting partial formatting adds a lot of complexity).
- **Opt-in** - All formatters are opt-in, it comes with no defaults.

The auto commands that handle format-on-save are also optional (you can disable them with `auto_commands = false`) and implement your own or wrap my implementation.

The core implementation is something I've been using for a long time, I just cleaned it up a bit and added non-LSP formatter support. I hope others might find it useful :)

**Comment by danielo515** [comment]
Formatting is just a part of what null-ls does
On Sun, 16 Jul 2023 at 13:19, David Elentok ***@***.***>
wrote:

> I've been using null-ls for a long time for both formatting and linting.
>
> For linting I replaced it with nvim-lint which seems to work pretty well
> so far.
>
> For formatting I had my own implementation for format-on-save that only
> used LSP formatting and non-LSP via null-ls, now that null-ls is archived I
> added direct support for formatting using non-LSP formatters and extracted
> the whole thing as a plugin -
> https://github.com/elentok/format-on-save.nvim
>
> I tried to keep it as simple as possible:
>
>    - *Synchronous formatting* - async formatting is very tricky since you
>    need to block the buffer from editing or deal with timing issues where the
>    buffer has been edited formatting. I tried using formatter.nvim but
>    couldn't get it to play well with LSP and the async thing didn't work well
>    for me.
>    - *Format on save only* - No partial formatting, only format entire
>    files (supporting partial formatting adds a lot of complexity).
>    - *Opt-in* - All formatters are opt-in, it comes with no defaults.
>
> The auto commands that handle format-on-save are also optional (you can
> disable them with auto_commands = false) and implement your own or wrap
> my implementation.
>
> The core implementation is something I've been using for a long time, I
> just cleaned it up a bit and added non-LSP formatter support. I hope others
> might find it useful :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LazyVim/LazyVim/discussions/1080#discussioncomment-6459997>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWNM3VJ3QPA2PZWPR63XQPE2TANCNFSM6AAAAAA2DYN4QQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
---
https://danielorodriguez.com

**Reply by elentok** [reply]
You're very right, sorry, didn't mean any disrespect, null-ls has been an integral part of my workflow and I'm very grateful for all the hard work 👍

**Comment by danielo515** [comment]
I never meant you were being disrespectful. I'm. Just to want to make sure
that people don't take a replacement that doesn't has all the features. If
required, maybe several replacements will be a viable solution

**Comment by robsonpeixoto** [comment]
Has someone tried [mattn/efm-langserver](https://github.com/mattn/efm-langserver)? Works well?

**Reply by xfzv** [reply]
> Has someone tried [mattn/efm-langserver](https://github.com/mattn/efm-langserver)? Works well?

I switched to [creativenull/efmls-configs-nvim](https://github.com/creativenull/efmls-configs-nvim) + [elentok/format-on-save.nvim](https://github.com/elentok/format-on-save.nvim), so far so good.

Also using [WhoIsSethDaniel/mason-tool-installer.nvim](https://github.com/WhoIsSethDaniel/mason-tool-installer.nvim) instead of [jay-babu/mason-null-ls.nvim](https://github.com/jay-babu/mason-null-ls.nvim).

**Comment by m0lson84** [comment]
To throw another package into the ring, I've stumbled upon [conform.nvim](https://github.com/stevearc/conform.nvim) which adds the formatter integration that `null-ls` provided. From the docs it seems to have a number of good influences and struck the right balance between convenience and extension.

**Comment by OneOfOne** [comment]
There's also [none-ls](https://github.com/nvimtools/none-ls.nvim) which is an active fork of null-ls, it's a drop in replacement for the most part.

https://github.com/nvimtools/none-ls.nvim

```lua
	{ 'jose-elias-alvarez/null-ls.nvim', enabled = false },
	{
		'nvimtools/none-ls.nvim',
		opts = function(_, opts)
			local nls = require('null-ls')  -- it's called null-ls internally
			opts.sources = vim.list_extend(opts.sources or {}, {
				nls.builtins.diagnostics.markdownlint,
				-- nls.builtins.completion.luasnip,
				nls.builtins.code_actions.gitsigns,
				-- go
				nls.builtins.code_actions.gomodifytags,
				nls.builtins.code_actions.impl,
				-- ts
				nls.builtins.formatting.biome,
				require("typescript.extensions.null-ls.code-actions")
			})
			return opts
		end,
	},

```

**Comment by fredrikaverpil** [comment]
[According to folke](https://github.com/LazyVim/LazyVim/discussions/1522#discussioncomment-7153402), the plan is to have [`none-ls`](https://github.com/nvimtools/none-ls.nvim) as the default, but maybe just for a couple of weeks.

```diff
return {
  {
-    "jose-elias-alvarez/null-ls.nvim",
+    "nvimtools/none-ls.nvim",
  },
}
```

**Comment by skinner12** [comment]
Is it easy to configure null-ls to none-ls?

Inviato da Outlook per iOS<https://aka.ms/o0ukef>
________________________________
Da: Fredrik Averpil ***@***.***>
Inviato: Saturday, September 30, 2023 11:30:41 PM
A: LazyVim/LazyVim ***@***.***>
Cc: RMazzu ***@***.***>; Author ***@***.***>
Oggetto: Re: [LazyVim/LazyVim] Archiving null-ls (Discussion #1080)


According to folke<https://github.com/LazyVim/LazyVim/discussions/1522#discussioncomment-7153402>, none-ls will be the default.

**Reply by folke** [reply]
just update LazyVim.
Old null-ls links will be automatically migrated to use none-ls instead.

**Comment by danielo515** [comment]
I'm OK with the legacy status of null-ls. Is there a simple and non
intrusive way I can stay on it?

**Reply by folke** [reply]
none-ls is the community fork of null-ls. IT's the same, just maintained

**Comment by danielo515** [comment]
Ok, thanks for the clarification.

**Comment by folke** [comment]
LazyVim 10.0 has just been released with removal of `none-ls`.

For more info, see https://github.com/LazyVim/LazyVim/blob/main/NEWS.md

**Reply by milch** [reply]
Curious why the change? none-ls seems to be active and receiving contributions

**Reply by folke** [reply]
In LazyVim we only used linters and formatters. none-ls is a bit overkill for just that.
You can of course still use none-ls if you want and if you use it's more advanced features.

**Reply by grimm26** [reply]
I think as mentioned previously here and by the null-ls maintainer, it is too tightly coupled with neovim and so takes a lot of effort to maintain as times goes on. That's my assumption for moving away. Honestly, I switched to nvim-lint + conform a few weeks ago and it is great. I especially like conform 👍

**Reply by OneOfOne** [reply]
Is there any alternatives for codeactions?

For example the go extra uses gomodifytags and impl, or gitsigns.

**Reply by EricDriussi** [reply]
That's exactly the issue I'm having as well: code actions.
For example, I have a piece of code that "adds a custom code action" that looks something like this:

```lua
null_ls.register({
	name = "more_actions",
	method = { require("null-ls").methods.CODE_ACTION },
	filetypes = { "_all" },
	generator = {
		fn = require("ts-node-action").available_actions,
	},
})
```

I'm not quite sure how to keep this functionality while migrating away from `null-ls`/`none-ls`.
I understand this is not a "help me pls" thread, it's just that I can see this being an issue for more people.

**Reply by dpetka2001** [reply]
You can still use `none-ls` along with `conform`. Can't you just keep your code actions in none-ls and just transition your formatting to conform?

**Reply by EricDriussi** [reply]
Yea, that shouldn't be an issue. It's just a matter of trimming down dependencies: I'll be swapping one plugin for two (which makes sense, `null-ls` had way to many responsibilities), but keeping `none-ls` just for code action integration seems a bit overkill.
If I can find a way to side-step `none-ls` for this use case I can ditch the plugin altogether.
Not that there's anything wrong with it, I'm just not sure how long it will last TBH.

**Reply by emmanueltouzery** [reply]
Regarding replacing the code actions parts of null-ls, I thought about making a telescope picker that would offer both true LSP actions, and custom actions as provided by the user. It could start from the existing telescope picker, just adding user provided code actions, without touching LSP at all. It actually doesn't seem too hard to do. But I'm personally unlikely to find the time.

**Comment by danielo515** [comment]
Maybe I am not reading you properly, but those are exactly the features
offered by null-ls, so it is just a replacement. Right?

**Comment by emmanueltouzery** [comment]
Null-ls is actually offering the features through the LSP API. What I meant was to "fake it". Offer a neovim gui front-end that allows to access the LSP features.. and  next to them actions that you write in lua. So it looks like what null-ls is offering, but the implementation is much simpler, I think, since you don't go through the LSP API or neovim LSP subsystem for the user-provided actions. But I guess it's off-topic for this issue.

**Comment by danielo515** [comment]
Ok, thanks for the clarification.

On Sun, Oct 15, 2023 at 8:58 PM emmanueltouzery ***@***.***>
wrote:

> Null-ls is actually offering the features through the LSP API. What I
> meant was to "fake it". Offer a neovim gui front-end that allows to access
> the LSP features.. and next to them actions that you write in lua. So it
> looks like what null-ls is offering, but the implementation is much
> simpler, I think, since you don't go through the LSP API or neovim LSP
> subsystem for the user-provided actions. But I guess it's off-topic for
> this issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LazyVim/LazyVim/discussions/1080#discussioncomment-7286410>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWLFE2GR777HE4WAW2TX7QW7BAVCNFSM6AAAAAA2DYN4QSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TEOBWGQYTA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com
